core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSailConnection.java

```diff
@@ -834,9 +907,10 @@ public void statementAdded(Statement statement) {
 
 	@Override
 	public void statementRemoved(Statement statement) {
-		if (preparedHasRun) {
+		if (prepareHasBeenCalled) {
 			throw new IllegalStateException("Detected changes after prepare() has been called.");
 		}
+		checkIfShapesRefreshIsNeeded(statement);
```

Comment:
I think this should preferably be called at the beginning of the prepare method, for better performance.